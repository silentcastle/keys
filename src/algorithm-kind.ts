import { InvalidKeyKindError } from "./invalid-key-kind.error";

export enum AlgorithmKind {
  secp256k1 = "secp256k1",
  ed25519 = "ed25519",
  x25519 = "x25519",
}

export namespace AlgorithmKind {
  export function fromString(s: string) {
    const maybe = s as AlgorithmKind;
    switch (maybe) {
      case AlgorithmKind.secp256k1:
        return AlgorithmKind.secp256k1;
      case AlgorithmKind.ed25519:
        return AlgorithmKind.ed25519;
      case AlgorithmKind.x25519:
        return AlgorithmKind.x25519;
      default:
        throw new InvalidKeyKindError(maybe);
    }
  }
}
