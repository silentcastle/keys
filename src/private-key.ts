import { KeyKind } from "./key-kind";
import { IPublicKey } from "./public-key";

export interface IPrivateKey {
  readonly kind: KeyKind;
  publicKey(): Promise<IPublicKey>;
}

export interface ISigner {
  kind: KeyKind;
  sign(message: Uint8Array): Promise<Uint8Array>;
}

export interface ISignerIdentified extends ISigner {
  kid: string;
}
