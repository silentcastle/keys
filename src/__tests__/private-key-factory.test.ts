import { PrivateKeyFactory } from "../private-key.factory";
import { AlgorithmKind } from "../algorithm-kind";

const privateKeyFactory = new PrivateKeyFactory();
const bytesSeed = new Uint8Array([1, 2, 3]);
const stringSeed = "hello";

describe("secp256k1", () => {
  test("generate from bytes", () => {
    const privateKey = privateKeyFactory.fromSeed(AlgorithmKind.secp256k1, bytesSeed);
    expect(privateKey.kind).toEqual(AlgorithmKind.secp256k1);
    expect(privateKey).toMatchSnapshot()
  });
  test("generate from string", () => {
    const privateKey = privateKeyFactory.fromSeed(AlgorithmKind.secp256k1, stringSeed);
    expect(privateKey.kind).toEqual(AlgorithmKind.secp256k1);
    expect(privateKey).toMatchSnapshot()
  });
});

describe("ed25519", () => {
  test("generate from bytes", () => {
    const privateKey = privateKeyFactory.fromSeed(AlgorithmKind.ed25519, bytesSeed);
    expect(privateKey.kind).toEqual(AlgorithmKind.ed25519);
    expect(privateKey).toMatchSnapshot()
  });
  test("generate from string", () => {
    const privateKey = privateKeyFactory.fromSeed(AlgorithmKind.ed25519, stringSeed);
    expect(privateKey.kind).toEqual(AlgorithmKind.ed25519);
    expect(privateKey).toMatchSnapshot()
  });
});
