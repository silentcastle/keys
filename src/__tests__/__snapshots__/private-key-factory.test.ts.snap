// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ed25519 generate from bytes 1`] = `
PrivateKey {
  "kind": "ed25519",
}
`;

exports[`ed25519 generate from string 1`] = `
PrivateKey {
  "kind": "ed25519",
}
`;

exports[`secp256k1 generate from bytes 1`] = `
PrivateKey {
  "kind": "secp256k1",
}
`;

exports[`secp256k1 generate from string 1`] = `
PrivateKey {
  "kind": "secp256k1",
}
`;
